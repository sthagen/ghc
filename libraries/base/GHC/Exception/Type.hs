{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE BangPatterns #-}
{-# OPTIONS_HADDOCK not-home #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Exception.Type
-- Copyright   :  (c) The University of Glasgow, 1998-2002
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- Exceptions and exception-handling functions.
--
-----------------------------------------------------------------------------

module GHC.Exception.Type
       ( Exception(..)    -- Class
       , SomeExceptionWithLocation(..), SomeException(..), ArithException(..)
        , addBacktrace
       , divZeroException, overflowException, ratioZeroDenomException
       , underflowException
       , traceException
       ) where

import Data.Maybe
import Data.Typeable (Typeable, cast)
   -- loop: Data.Typeable -> GHC.Err -> GHC.Exception
import GHC.Base
import GHC.Show
import GHC.Exception.Backtrace (Backtrace)

-- TODO: Remove!
import {-# SOURCE #-} Debug.Trace

{- |
The @SomeExceptionWithLocation@ type is the root of the exception type hierarchy.
When an exception of type @e@ is thrown, behind the scenes it is
encapsulated in a @SomeException@ with is wrapped by @SomeExceptionWithLocation@.
-}
data SomeExceptionWithLocation = SomeExceptionWithLocation !SomeException ![Backtrace]

data SomeException = forall e . Exception e => SomeException !e

-- | @since 3.0
instance Show SomeExceptionWithLocation where
    -- TODO: Print backtraces
    showsPrec p (SomeExceptionWithLocation (SomeException e) _) = showsPrec p e

instance Show SomeException where
    -- TODO: Fix show
    showsPrec p (SomeException e) = showsPrec p e

-- | Add a 'Backtrace' to the list of backtraces.
--
-- @since 4.16.0.0
addBacktrace :: Backtrace -> SomeExceptionWithLocation -> SomeExceptionWithLocation
addBacktrace bt (SomeExceptionWithLocation e bts) =
    SomeExceptionWithLocation e (bt : bts)

{- |
Any type that you wish to throw or catch as an exception must be an
instance of the @Exception@ class. The simplest case is a new exception
type directly below the root:

> data MyException = ThisException | ThatException
>     deriving Show
>
> instance Exception MyException

The default method definitions in the @Exception@ class do what we need
in this case. You can now throw and catch @ThisException@ and
@ThatException@ as exceptions:

@
*Main> throw ThisException \`catch\` \\e -> putStrLn (\"Caught \" ++ show (e :: MyException))
Caught ThisException
@

In more complicated examples, you may wish to define a whole hierarchy
of exceptions:

> ---------------------------------------------------------------------
> -- Make the root exception type for all the exceptions in a compiler
>
> data SomeCompilerException = forall e . Exception e => SomeCompilerException e
>
> instance Show SomeCompilerException where
>     show (SomeCompilerException e) = show e
>
> instance Exception SomeCompilerException
>
> compilerExceptionToException :: Exception e => e -> SomeExceptionWithLocation
> compilerExceptionToException = toException . SomeCompilerException
>
> compilerExceptionFromException :: Exception e => SomeExceptionWithLocation -> Maybe e
> compilerExceptionFromException x = do
>     SomeCompilerException a <- fromException x
>     cast a
>
> ---------------------------------------------------------------------
> -- Make a subhierarchy for exceptions in the frontend of the compiler
>
> data SomeFrontendException = forall e . Exception e => SomeFrontendException e
>
> instance Show SomeFrontendException where
>     show (SomeFrontendException e) = show e
>
> instance Exception SomeFrontendException where
>     toException = compilerExceptionToException
>     fromException = compilerExceptionFromException
>
> frontendExceptionToException :: Exception e => e -> SomeExceptionWithLocation
> frontendExceptionToException = toException . SomeFrontendException
>
> frontendExceptionFromException :: Exception e => SomeExceptionWithLocation -> Maybe e
> frontendExceptionFromException x = do
>     SomeFrontendException a <- fromException x
>     cast a
>
> ---------------------------------------------------------------------
> -- Make an exception type for a particular frontend compiler exception
>
> data MismatchedParentheses = MismatchedParentheses
>     deriving Show
>
> instance Exception MismatchedParentheses where
>     toException   = frontendExceptionToException
>     fromException = frontendExceptionFromException

We can now catch a @MismatchedParentheses@ exception as
@MismatchedParentheses@, @SomeFrontendException@ or
@SomeCompilerException@, but not other types, e.g. @IOException@:

@
*Main> throw MismatchedParentheses \`catch\` \\e -> putStrLn (\"Caught \" ++ show (e :: MismatchedParentheses))
Caught MismatchedParentheses
*Main> throw MismatchedParentheses \`catch\` \\e -> putStrLn (\"Caught \" ++ show (e :: SomeFrontendException))
Caught MismatchedParentheses
*Main> throw MismatchedParentheses \`catch\` \\e -> putStrLn (\"Caught \" ++ show (e :: SomeCompilerException))
Caught MismatchedParentheses
*Main> throw MismatchedParentheses \`catch\` \\e -> putStrLn (\"Caught \" ++ show (e :: IOException))
*** Exception: MismatchedParentheses
@

-}
class (Typeable e, Show e) => Exception e where
    toException   :: e -> SomeExceptionWithLocation
    fromException :: SomeExceptionWithLocation -> Maybe e

    toException e = SomeExceptionWithLocation (SomeException e) []
    fromException (SomeExceptionWithLocation (SomeException e) _) = cast e

    -- | Render this exception value in a human-friendly manner.
    --
    -- Default implementation: @'show'@.
    --
    -- @since 4.8.0.0
    displayException :: e -> String
    displayException = show

    toString :: e -> String
    toString e = show e

instance Exception SomeException where
  toException e = SomeExceptionWithLocation e []
  fromException (SomeExceptionWithLocation e _) = Just e
  toString (SomeException e) = "SomeException " ++ toString e

instance Exception SomeExceptionWithLocation where
    toException se = se
    fromException = Just
    displayException (SomeExceptionWithLocation e _) = displayException e
    toString (SomeExceptionWithLocation e _) = "SomeExceptionWithLocation " ++ toString e

traceException :: Exception e => e -> e
traceException e = trace (toString e) e

-- |Arithmetic exceptions.
data ArithException
  = Overflow
  | Underflow
  | LossOfPrecision
  | DivideByZero
  | Denormal
  | RatioZeroDenominator -- ^ @since 4.6.0.0
  deriving ( Eq  -- ^ @since 3.0
           , Ord -- ^ @since 3.0
           )

divZeroException, overflowException, ratioZeroDenomException, underflowException  :: SomeExceptionWithLocation
divZeroException        = toException DivideByZero
overflowException       = toException Overflow
ratioZeroDenomException = toException RatioZeroDenominator
underflowException      = toException Underflow

-- | @since 4.0.0.0
instance Exception ArithException

-- | @since 4.0.0.0
instance Show ArithException where
  showsPrec _ Overflow        = showString "arithmetic overflow"
  showsPrec _ Underflow       = showString "arithmetic underflow"
  showsPrec _ LossOfPrecision = showString "loss of precision"
  showsPrec _ DivideByZero    = showString "divide by zero"
  showsPrec _ Denormal        = showString "denormal"
  showsPrec _ RatioZeroDenominator = showString "Ratio has zero denominator"
