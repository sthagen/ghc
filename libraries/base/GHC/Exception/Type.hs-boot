{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Exception.Type
  ( SomeExceptionWithLocation
  , divZeroException
  , overflowException
  , ratioZeroDenomException
  , underflowException
  , Exception
  ) where

import GHC.Num.Integer ()   -- See Note [Depend on GHC.Num.Integer] in GHC.Base

data SomeExceptionWithLocation
divZeroException, overflowException,
  ratioZeroDenomException, underflowException :: SomeExceptionWithLocation

class Exception e

instance Exception SomeExceptionWithLocation
