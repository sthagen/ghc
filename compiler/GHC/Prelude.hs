{-# LANGUAGE CPP #-}
{-# OPTIONS_HADDOCK not-home #-}
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE ViewPatterns #-}

-- | Custom GHC "Prelude"
--
-- This module serves as a replacement for the "Prelude" module
-- and abstracts over differences between the bootstrapping
-- GHC version, and may also provide a common default vocabulary.

-- Every module in GHC
--   * Is compiled with -XNoImplicitPrelude
--   * Explicitly imports GHC.Prelude

module GHC.Prelude
  (module X
  ,module Bits
  ,shiftL, shiftR
#if __GLASGOW_HASKELL__ < 903
  ,SomeExceptionWithLocation
  ,pattern SomeExceptionWithLocation
#else
  ,SomeExceptionWithLocation(..)
#endif
  ) where


-- We export the 'Semigroup' class but w/o the (<>) operator to avoid
-- clashing with the (Outputable.<>) operator which is heavily used
-- through GHC's code-base.

{-
Note [Why do we import Prelude here?]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The files ghc-boot-th.cabal, ghc-boot.cabal, ghci.cabal and
ghc-heap.cabal contain the directive default-extensions:
NoImplicitPrelude. There are two motivations for this:
  - Consistency with the compiler directory, which enables
    NoImplicitPrelude;
  - Allows loading the above dependent packages with ghc-in-ghci,
    giving a smoother development experience when adding new
    extensions.
-}

import Prelude as X hiding ((<>))
import Data.Foldable as X (foldl')
#if __GLASGOW_HASKELL__ < 903
import Control.Exception ( SomeException(..) )
#else
import Control.Exception ( SomeExceptionWithLocation(..) )
#endif

#if MIN_VERSION_base(4,16,0)
import GHC.Bits as Bits hiding (shiftL, shiftR)
# if defined(DEBUG)
import qualified GHC.Bits as Bits (shiftL, shiftR)
# endif

#else
--base <4.15
import Data.Bits as Bits hiding (shiftL, shiftR)
# if defined(DEBUG)
import qualified Data.Bits as Bits (shiftL, shiftR)
# endif
#endif

{- Note [Default to unsafe shifts inside GHC]
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The safe shifts can introduce branches which come
at the cost of performance. We still want the additional
debugability for debug builds. So we define it as one or the
other depending on the DEBUG setting.

Why do we then continue on to re-export the rest of Data.Bits?
If we would not what is likely to happen is:
* Someone imports Data.Bits, uses xor. Things are fine.
* They add a shift and get an ambigious definition error.
* The are puzzled for a bit.
* They either:
  + Remove the import of Data.Bits and get an error because xor is not in scope.
  + Add the hiding clause to the Data.Bits import for the shifts.

Either is quite annoying. Simply re-exporting all of Data.Bits avoids this
making for a smoother developer experience. At the cost of having a few more
names in scope at all time. But that seems like a fair tradeoff.

See also #19618
-}

-- We always want the Data.Bits method to show up for rules etc.
{-# INLINE shiftL #-}
{-# INLINE shiftR #-}
shiftL, shiftR :: Bits.Bits a => a -> Int -> a
#if defined(DEBUG)
shiftL = Bits.shiftL
shiftR = Bits.shiftR
#else
shiftL = Bits.unsafeShiftL
shiftR = Bits.unsafeShiftR
#endif

#if __GLASGOW_HASKELL__ < 903
type SomeExceptionWithLocation = SomeException

{-# COMPLETE SomeExceptionWithLocation #-}
pattern SomeExceptionWithLocation :: forall. SomeException -> () -> SomeException
pattern SomeExceptionWithLocation e unit <- (\x -> ((), x) -> (unit, e))
  where
    SomeExceptionWithLocation (SomeException e) _ = SomeException e
#endif
