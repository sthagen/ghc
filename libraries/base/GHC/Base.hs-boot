{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Base (Maybe, Semigroup, Monoid, String) where

import GHC.Maybe (Maybe)
import GHC.Types (Char)

class Semigroup a
class Monoid a

type String = [Char]
