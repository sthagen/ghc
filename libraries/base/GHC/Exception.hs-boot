{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE ExistentialQuantification#-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TypeInType #-}

{-
This SOURCE-imported hs-boot module cuts a big dependency loop:

         GHC.Exception
imports  Data.Maybe
imports  GHC.Base
imports  GHC.Err
imports  {-# SOURCE #-} GHC.Exception

More dramatically

         GHC.Exception
imports  Data.Typeable
imports  Data.Typeable.Internal
imports  GHC.Arr (fingerprint representation etc)
imports  GHC.Real
imports  {-# SOURCE #-} GHC.Exception

However, GHC.Exceptions loop-breaking exports are all nice,
well-behaved, non-bottom values.  The clients use 'raise#'
to get a visibly-bottom value.
-}

module GHC.Exception
  ( module GHC.Exception.Type
  , errorCallException
  , errorCallWithCallStackException
  , throw
  ) where

import {-# SOURCE #-} GHC.Exception.Type
import GHC.Types ( Char, RuntimeRep, TYPE )
import GHC.Stack.Types ( CallStack, HasCallStack )

errorCallException :: [Char] -> SomeExceptionWithLocation
errorCallWithCallStackException :: [Char] -> CallStack -> SomeExceptionWithLocation

throw :: HasCallStack => forall (r :: RuntimeRep). forall (a :: TYPE r). forall e. Exception e => e -> a
