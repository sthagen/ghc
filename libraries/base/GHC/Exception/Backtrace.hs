{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# OPTIONS_HADDOCK not-home #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Exception.Backtrace
-- Copyright   :  (c) The GHC Team, 2020-2021
-- Authors     :  Ben Gamari, David Eichmann, Sven Tennie
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- Collect Exception backtraces with several mechanisms.
-----------------------------------------------------------------------------

module GHC.Exception.Backtrace
  ( Backtrace (..),
    BacktraceMechanism (..),
    setDefaultBacktraceMechanisms,
    getDefaultBacktraceMechanisms,
    showBacktraces,
    collectBacktraces,
  )
where

import Data.List
import Data.Maybe
import GHC.Base
import {-# SOURCE #-} GHC.ExecutionStack (Location, getStackTrace)
import {-# SOURCE #-} GHC.ExecutionStack.Internal (showStackFrames)
import GHC.IO.Unsafe
import {-# SOURCE #-} GHC.IORef
import GHC.Ptr
import GHC.Show
import {-# SOURCE #-} GHC.Stack
import {-# SOURCE #-} GHC.Stack.CCS
import {-# SOURCE #-} GHC.Stack.CloneStack

-- | An exception backtrace.
--
-- @since 4.15
data Backtrace
  = -- | a cost center profiler backtrace
    CostCenterBacktrace (Ptr GHC.Stack.CCS.CostCentreStack)
  | -- | a stack from 'GHC.Stack.HasCallStack'
    HasCallStackBacktrace GHC.Stack.CallStack
  | -- | a stack unwinding (e.g. DWARF) backtrace
    ExecutionBacktrace [GHC.ExecutionStack.Location]
  | -- | a backtrace from Info Table Provenance Entries
    IPEBacktrace [StackEntry]

-- | @since 4.15
instance Show Backtrace where
  -- TODO
  showsPrec p (CostCenterBacktrace ccs) = showsPrec p ccs
  showsPrec p (HasCallStackBacktrace ccs) = showsPrec p ccs
  showsPrec _ (ExecutionBacktrace ccs) = showStackFrames ccs
  showsPrec p (IPEBacktrace stackEntries) = showsPrec p stackEntries

-- | How to collect a backtrace when an exception is thrown.
data BacktraceMechanism
  = -- | collect a cost center stacktrace (only available when built with profiling)
    CostCenterBacktraceMech
  | -- | use execution stack unwinding with given limit
    ExecutionStackBacktraceMech (Maybe Int)
  | -- | collect backtraces from Info Table Provenance Entries
    IPEBacktraceMech
  deriving (Eq, Show)

showBacktraces :: [Backtrace] -> String
showBacktraces bts = unlines $ intersperse "" $ map show bts

currentBacktraceMechanisms :: IORef [BacktraceMechanism]
currentBacktraceMechanisms = unsafePerformIO $ newIORef []
{-# NOINLINE currentBacktraceMechanisms #-}

-- | Set how 'Control.Exception.throwIO', et al. collect backtraces.
setDefaultBacktraceMechanisms :: [BacktraceMechanism] -> IO ()
setDefaultBacktraceMechanisms = writeIORef currentBacktraceMechanisms

-- | Returns the currently selected 'BacktraceMechanism'.
getDefaultBacktraceMechanisms :: IO [BacktraceMechanism]
getDefaultBacktraceMechanisms = readIORef currentBacktraceMechanisms

-- | Collect a list of 'Backtrace' via all current default 'BacktraceMechanism'.
-- See 'setDefaultBacktraceMechanisms'
collectBacktraces :: IO [Backtrace]
collectBacktraces = do
  mech <- getDefaultBacktraceMechanisms
  catMaybes `fmap` mapM collectBacktraces' mech

-- | Collect a 'Backtrace' via the given 'BacktraceMechanism'.
collectBacktraces' :: BacktraceMechanism -> IO (Maybe Backtrace)
collectBacktraces' CostCenterBacktraceMech = do
  ptr <- getCurrentCCS ()
  return $ if ptr == nullPtr then Nothing else Just (CostCenterBacktrace ptr)
collectBacktraces' (ExecutionStackBacktraceMech _) = fmap ExecutionBacktrace `fmap` getStackTrace
collectBacktraces' IPEBacktraceMech = do
  stack <- cloneMyStack
  stackEntries <- decode stack
  return $ Just $ IPEBacktrace stackEntries
