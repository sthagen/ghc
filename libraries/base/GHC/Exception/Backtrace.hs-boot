{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Exception.Backtrace
    ( Backtrace
    , showBacktraces
    , collectBacktraces
    ) where

import GHC.Base

data Backtrace

showBacktraces :: [Backtrace] -> String

collectBacktraces :: IO [Backtrace]
