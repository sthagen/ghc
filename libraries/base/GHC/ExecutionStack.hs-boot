{-# LANGUAGE NoImplicitPrelude #-}

module GHC.ExecutionStack
  ( Location,
    getStackTrace,
  )
where

import GHC.Base
import {-# SOURCE #-} GHC.ExecutionStack.Internal (Location)

getStackTrace :: IO (Maybe [Location])
