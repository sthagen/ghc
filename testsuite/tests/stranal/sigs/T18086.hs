{-# OPTIONS_GHC -O2 -fforce-recomp #-}
module T18086 where

import GHC.Stack
import GHC.Utils.Panic.Plain
import Control.Exception
import System.IO.Unsafe

-- Should have strictness signature <L>x, emphasis on the exceptional
-- divergence result.
m :: IO ()
m = do
  putStrLn "foo"
  error "bar"

-- Ditto, just in a more complex scenario (the original reproducer of #18086)
-- NB: This depends on the strictness signature of GHC.Magic.runRW#, which will
-- be topSig if compiled with optimisation as in the quick flavour. In this case
-- we'll see <L>, without x.
panic :: String -> a
panic x = unsafeDupablePerformIO $ do
  stack <- ccsToStrings =<< getCurrentCCS x
  if null stack
  then throw (PlainPanic x)
  else throw (PlainPanic (x ++ '\n' : renderStack stack))

