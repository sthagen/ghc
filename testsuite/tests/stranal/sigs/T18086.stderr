
==================== Strictness signatures ====================
T18086.$trModule:
T18086.m: <SL>x
T18086.panic: <L>x



==================== Cpr signatures ====================
T18086.$trModule:
T18086.m: b
T18086.panic: b



==================== Strictness signatures ====================
T18086.$trModule:
T18086.m: <SL>x
T18086.panic: <L>x


